### PR TITLE
Support and/or section

### PR DESCRIPTION
**Pull request by okkez**
For example:

    <filter>
      @type grep
      <and>
        <regexp>
          key level
          pattern ^WARN|ERROR$
        </regexp>
        <regexp>
          key method
          pattern ^GET$
        </regexp>
      </and>
    </filter>

For more details, see test code

See #1858

### CODE REVIEW

File: lib/fluent/plugin/filter_grep.rb [hunk 1]
```diff
@@ -25,10 +25,14 @@ class GrepFilter < Filter
...
 
-      @_regexps = {}
-      @_excludes = {}
+      @_regexp_and_conditions = {}
+      @_exclude_and_conditions = {}
+      @_regexp_or_conditions = {}
+      @_exclude_or_conditions = {}
     end
 
+    attr_reader :_regexp_and_conditions, :_exluce_and_conditions, :_regexp_or_conditions, :_exclude_or_conditions
```
Comment (JohnStarich) [inline lib/fluent/plugin/filter_grep.rb]: small spelling mistake: `_exluce_and_conditions` -> `_exclude_and_conditions`

Comment (okkez) [reply lib/fluent/plugin/filter_grep.rb]: Thanks! I've fixed.

File: lib/fluent/plugin/filter_grep.rb [hunk 2]
```diff
@@ -122,5 +201,18 @@ def filter(tag, time, record)
       end
       result
     end
+
+    class Expression
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_grep.rb]: How about using Struct for it?

```
Expression = Struct.new(:key, :pattern) do
  def match?(record)
    ::Fluent::StringUtil.match_regexp(pattern, key.call(record).to_s)
  end
end
```

File: lib/fluent/plugin/filter_grep.rb [hunk 3]
```diff
@@ -54,65 +59,139 @@ def initialize
...
       end
     end
 
     def filter(tag, time, record)
       result = nil
       begin
         catch(:break_loop) do
-          @_regexps.each do |key, regexp|
-            throw :break_loop unless ::Fluent::StringUtil.match_regexp(regexp, key.call(record).to_s)
+          @_regexp_and_conditions.each do |key, expression|
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_grep.rb]: This is very small optimization but how about using array for conditions?
In `filter`, we don't use `key` and hash related operations so array is good structure.

File: lib/fluent/plugin/filter_grep.rb [hunk 4]
```diff
@@ -54,65 +59,139 @@ def initialize
...
       super
 
-      rs = {}
       (1..REGEXP_MAX_NUM).each do |i|
         next unless conf["regexp#{i}"]
         key, regexp = conf["regexp#{i}"].split(/ /, 2)
         raise Fluent::ConfigError, "regexp#{i} does not contain 2 parameters" unless regexp
-        raise Fluent::ConfigError, "regexp#{i} contains a duplicated key, #{key}" if rs[key]
-        rs[key] = Regexp.compile(regexp)
+        raise Fluent::ConfigError, "regexp#{i} contains a duplicated key, #{key}" if @_regexp_and_conditions[key]
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_grep.rb]: How about logging info or warning message for top level setting?
`Top level multiple <regexp> is intepreted as 'and' condition`.

**Comment by JohnStarich** [comment]
This looks great. @repeatedly can this get merged? (The failing test in Travis looks unrelated to these changes)

**Comment by repeatedly** [comment]
Sorry for the delay. Will check it again!

File: lib/fluent/plugin/filter_grep.rb [hunk 5]
```diff
@@ -54,65 +59,157 @@ def initialize
...
-    # for test
-    attr_reader :_regexps
-    attr_reader :_excludes
+    config_section :and, param_name: :and_conditions, multi: true do
+      config_section :regexp, param_name: :regexps, multi: true do
+        desc "The field name to which the regular expression is applied."
+        config_param :key, :string
+        desc "The regular expression."
+        config_param :pattern do |value|
+          if value.start_with?("/") and value.end_with?("/")
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_grep.rb]: Hmm... I see these patterns, we need to `:regexp` type for config_param.

@okkez Could you write a patch for adding `regexp` type in other PR?

Comment (okkez) [reply lib/fluent/plugin/filter_grep.rb]: Sure, I will write a patch for this.

File: lib/fluent/plugin/filter_grep.rb [hunk 6]
```diff
@@ -54,65 +59,157 @@ def initialize
...
+          end
+        end
+      end
+    end
 
     def configure(conf)
       super
 
-      rs = {}
+      _regexp_and_conditions = {}
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_grep.rb]: This name is conflicts with attr_reader's definition. Yeah, I know local variable is preferred but using other name , e.g. `_regexp_and_conds`, is better. Others are same.

File: lib/fluent/plugin/filter_grep.rb (hunk 5 above)
Comment (okkez) [reply lib/fluent/plugin/filter_grep.rb]: I've created #1927

**Comment by repeatedly** [comment]
@okkez We enabled DCO for fluentd repository. Could you re-commit last commit with `-s`?

**Comment by okkez** [comment]
I've re-committed and pushed the last commit with `-s`. @repeatedly

**Comment by repeatedly** [comment]
Thanks!

**Pull request merged**